### ISSUE TITLE
Incorrect Padding Error in pdf-report.gmp.py file

### ISSUE THREAD

**Issue by BhagyashreeKoyande** [greenbone/gvm-tools#327]
When I run the pdf-report.gmp.py through gvm-script command, I get an error as Incorrect Padding.

![IMG-20201105-WA0010](https://user-images.githubusercontent.com/42887114/98224246-455a9400-1f79-11eb-97fb-ab520dc2b511.jpg)

**Comment by y0urself** [comment]
You may take time to give more information? There is an Issue-Template you might have seen. What version of gvm-tools and python-gvm did you used?

**Comment by BhagyashreeKoyande** [comment]
This is the code I am using and you can see the error in screenshot attached at first

# -*- coding: utf-8 -*-
# Copyright (C) 2019 Greenbone Networks GmbH
#
# SPDX-License-Identifier: GPL-3.0-or-later
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from base64 import b64decode
from pathlib import Path


def check_args(args):
    len_args = len(args.script) - 1
    if len_args < 1:
        message = """
        This script requests the given report and saves it as a pdf file locally.
        It needs one parameters after the script name.
        1. <report_id>     -- ID of the report
        
        Optional a file name to save the pdf in.
        Example:
            $ gvm-script --gmp-username name --gmp-password pass \
ssh --hostname <gsm> scripts/pdf-report.gmp.py <report_id> <pdf_file>
        """
        print(message)
        quit()


def main(gmp, args):
    # check if report id and PDF filename are provided to the script
    # argv[0] contains the script name
    check_args(args)

    report_id = args.argv[1]
    if len(args.argv) == 3:
        pdf_filename = args.argv[2]
    else:
        pdf_filename = args.argv[1] + ".pdf"

    pdf_report_format_id = "c402cc3e-b531-11e1-9163-406186ea4fc5"

    response = gmp.get_report(
        report_id=report_id, report_format_id=pdf_report_format_id
    )

    report_element = response.find("report")
    # get the full content of the report element
    content = report_element.find("report_format").tail

    # convert content to 8-bit ASCII bytes
    binary_base64_encoded_pdf = content.encode('ascii')

    # decode base64
    binary_pdf = b64decode(binary_base64_encoded_pdf)

    # write to file and support ~ in filename path
    pdf_path = Path(pdf_filename).expanduser()

    pdf_path.write_bytes(binary_pdf)

    print('Done. PDF created: ' + str(pdf_path))


if __name__ == '__gmp__':
    main(gmp, args)

**Comment by BhagyashreeKoyande** [comment]
I have clone the git directory for files and installed gvm-tools using pip3.

**Comment by y0urself** [comment]
you may type `gvm-script -V` in the console and copy the output here?

**Comment by BhagyashreeKoyande** [comment]
gvm-script 20.10.1 (API version 20.11.0)

**Comment by y0urself** [comment]
Hmm interesting. I don't have any trouble with this script:
```
gvm-scr 192.168.178.138 ../gvm-tools/scripts/pdf-report.gmp.py 6373f0ac-9ebc-4668-a44d-057318addf88
Done. PDF created: 6373f0ac-9ebc-4668-a44d-057318addf88.pdf
```

**Comment by BhagyashreeKoyande** [comment]
Ohh, any idea what is wrong with my gvm-script command?

**Comment by BhagyashreeKoyande** [comment]
gvm-script scan command works for me and generates report id also. Problems occurs while generating pdf report. I am also attaching the screenshot of terminal output
![IMG-20201105-WA0012](https://user-images.githubusercontent.com/42887114/98226282-d0d52480-1f7b-11eb-9464-041f6c54f332.jpg)


Please check and guide me for what I can do

**Comment by y0urself** [comment]
Maybe it is the report?
If you want, you may try to download the report as xml?
```
gvm-pyshell --gmp-username=admin --gmp-password=admin socket <your_socket>
>>> from gvm.xml import pretty_print
>>> response = gmp.get_report(report_id='<your_report_id>', report_format_id='a994b278-1f62-11e1-96ac-406186ea4fc5')
>>> with open('<any_file>', 'w') as f:
...     pretty_print(response, f)
...
```
Is this working? Is the file `<any_file>` containing a valid xml Report?

**Comment by bjoernricks** [comment]
The error indicates that the returned content (the XML response from our `gvmd` daemon) didn't contain valid base64 encoded data. Therefore we need take a to look at XML output of the response for being able to debug the issue further. Please follow the advise from @y0urself and paste the formatted XML here.

**Comment by BhagyashreeKoyande** [comment]
Yes, I followed the steps as instructed by @y0urself . The file contains xml response and the content is as follows:

<get_reports_response status="200" status_text="OK">
  <report id="03948874-9f06-4231-ba43-04dc3c778240" format_id="a994b278-1f62-11e1-96ac-406186ea4fc5" extension="xml" content_type="text/xml">
    <owner>
      <name>admin</name>
    </owner>
    <name/>
    <comment/>
    <creation_time/>
    <modification_time>2020-11-05T09:54:48Z</modification_time>
    <writable>0</writable>
    <in_use>0</in_use>
    <task id="40b6c517-ba10-4fbc-811c-1435c28be1b2">
      <name>Scan Suspect Host 192.168.56.1</name>
    </task>
    <report_format id="a994b278-1f62-11e1-96ac-406186ea4fc5">
      <name>XML</name>
    </report_format>
    <report id="03948874-9f06-4231-ba43-04dc3c778240">
      <gmp>
        <version>9.0</version>
      </gmp>
      <sort>
        <field>name<order>ascending</order></field>
      </sort>
      <filters id="">
        <term>apply_overrides=0 min_qod=70 first=1 rows=1000 sort=name</term>
        <filter>High</filter>
        <filter>Medium</filter>
        <filter>Low</filter>
        <filter>Log</filter>
        <filter>Debug</filter>
        <keywords>
          <keyword>
            <column>apply_overrides</column>
            <relation>=</relation>
            <value>0</value>
          </keyword>
          <keyword>
            <column>min_qod</column>
            <relation>=</relation>
            <value>70</value>
          </keyword>
          <keyword>
            <column>first</column>
            <relation>=</relation>
            <value>1</value>
          </keyword>
          <keyword>
            <column>rows</column>
            <relation>=</relation>
            <value>1000</value>
          </keyword>
          <keyword>
            <column>sort</column>
            <relation>=</relation>
            <value>name</value>
          </keyword>
        </keywords>
      </filters>
      <severity_class id="d4c74cda-89e1-11e3-9c29-406186ea4fc5">
        <name>nist</name>
        <full_name>NVD Vulnerability Severity Ratings</full_name>
        <severity_range>
          <name>None</name>
          <min>0.0</min>
          <max>0.0</max>
        </severity_range>
        <severity_range>
          <name>Low</name>
          <min>0.1</min>
          <max>3.9</max>
        </severity_range>
        <severity_range>
          <name>Medium</name>
          <min>4.0</min>
          <max>6.9</max>
        </severity_range>
        <severity_range>
          <name>High</name>
          <min>7.0</min>
          <max>10.0</max>
        </severity_range>
      </severity_class>
      <scan_run_status>Done</scan_run_status>
      <hosts>
        <count>0</count>
      </hosts>
      <closed_cves>
        <count>0</count>
      </closed_cves>
      <vulns>
        <count>0</count>
      </vulns>
      <os>
        <count>0</count>
      </os>
      <apps>
        <count>0</count>
      </apps>
      <ssl_certs>
        <count>0</count>
      </ssl_certs>
      <task id="40b6c517-ba10-4fbc-811c-1435c28be1b2">
        <name>Scan Suspect Host 192.168.56.1</name>
        <comment/>
        <target id="3006db49-89c2-435b-9d52-77e0ad414f4b">
          <trash>0</trash>
          <name>Suspect Host 192.168.56.1 2020-11-05 15:24:48.003377</name>
          <comment/>
        </target>
        <progress>-1</progress>
      </task>
      <scan>
        <task/>
      </scan>
      <timestamp>2020-11-05T09:54:48Z</timestamp>
      <scan_start/>
      <timezone>Coordinated Universal Time</timezone>
      <timezone_abbrev>UTC</timezone_abbrev>
      <result_count>0<full>0</full><filtered>0</filtered><debug><full>0</full><filtered>0</filtered></debug><hole><full>0</full><filtered>0</filtered></hole><info><full>0</full><filtered>0</filtered></info><log><full>0</full><filtered>0</filtered></log><warning><full>0</full><filtered>0</filtered></warning><false_positive><full>0</full><filtered>0</filtered></false_positive></result_count>
      <severity>
        <full>-3.0</full>
        <filtered>-3.0</filtered>
      </severity>
      <scan_end>2020-11-05T09:54:48Z</scan_end>
      <errors>
        <count>1</count>
        <error>
          <host>
            <asset asset_id=""/>
          </host>
          <port/>
          <description>Could not connect to Scanner</description>
          <nvt oid="">
            <type>nvt</type>
            <name/>
            <cvss_base/>
          </nvt>
          <scan_nvt_version/>
          <severity>-3</severity>
        </error>
      </errors>
      <report_format/>
    </report>
  </report>
  <filters id="">
    <term>apply_overrides=0 min_qod=70 first=1 rows=10 sort=name</term>
    <keywords>
      <keyword>
        <column>apply_overrides</column>
        <relation>=</relation>
        <value>0</value>
      </keyword>
      <keyword>
        <column>min_qod</column>
        <relation>=</relation>
        <value>70</value>
      </keyword>
      <keyword>
        <column>first</column>
        <relation>=</relation>
        <value>1</value>
      </keyword>
      <keyword>
        <column>rows</column>
        <relation>=</relation>
        <value>10</value>
      </keyword>
      <keyword>
        <column>sort</column>
        <relation>=</relation>
        <value>name</value>
      </keyword>
    </keywords>
  </filters>
  <sort>
    <field>name<order>ascending</order></field>
  </sort>
  <reports start="1" max="1000"/>
  <report_count>8<filtered>1</filtered><page>1</page></report_count>
</get_reports_response>

**Comment by BhagyashreeKoyande** [comment]
Yes, I followed the steps as instructed by @y0urself . The file contains xml response and the content is as follows:
[demo.txt](https://github.com/greenbone/gvm-tools/files/5493577/demo.txt)

**Comment by bjoernricks** [comment]
I've you paste thing please always use [preformatted markdown syntax](https://docs.github.com/en/free-pro-team@latest/github/writing-on-github/creating-and-highlighting-code-blocks)

**Comment by BhagyashreeKoyande** [comment]
```
<get_reports_response status="200" status_text="OK">
  <report id="03948874-9f06-4231-ba43-04dc3c778240" format_id="a994b278-1f62-11e1-96ac-406186ea4fc5" extension="xml" content_type="text/xml">
    <owner>
      <name>admin</name>
    </owner>
    <name/>
    <comment/>
    <creation_time/>
    <modification_time>2020-11-05T09:54:48Z</modification_time>
    <writable>0</writable>
    <in_use>0</in_use>
    <task id="40b6c517-ba10-4fbc-811c-1435c28be1b2">
      <name>Scan Suspect Host 192.168.56.1</name>
    </task>
    <report_format id="a994b278-1f62-11e1-96ac-406186ea4fc5">
      <name>XML</name>
    </report_format>
    <report id="03948874-9f06-4231-ba43-04dc3c778240">
      <gmp>
        <version>9.0</version>
      </gmp>
      <sort>
        <field>name<order>ascending</order></field>
      </sort>
      <filters id="">
        <term>apply_overrides=0 min_qod=70 first=1 rows=1000 sort=name</term>
        <filter>High</filter>
        <filter>Medium</filter>
        <filter>Low</filter>
        <filter>Log</filter>
        <filter>Debug</filter>
        <keywords>
          <keyword>
            <column>apply_overrides</column>
            <relation>=</relation>
            <value>0</value>
          </keyword>
          <keyword>
            <column>min_qod</column>
            <relation>=</relation>
            <value>70</value>
          </keyword>
          <keyword>
            <column>first</column>
            <relation>=</relation>
            <value>1</value>
          </keyword>
          <keyword>
            <column>rows</column>
            <relation>=</relation>
            <value>1000</value>
          </keyword>
          <keyword>
            <column>sort</column>
            <relation>=</relation>
            <value>name</value>
          </keyword>
        </keywords>
      </filters>
      <severity_class id="d4c74cda-89e1-11e3-9c29-406186ea4fc5">
        <name>nist</name>
        <full_name>NVD Vulnerability Severity Ratings</full_name>
        <severity_range>
          <name>None</name>
          <min>0.0</min>
          <max>0.0</max>
        </severity_range>
        <severity_range>
          <name>Low</name>
          <min>0.1</min>
          <max>3.9</max>
        </severity_range>
        <severity_range>
          <name>Medium</name>
          <min>4.0</min>
          <max>6.9</max>
        </severity_range>
        <severity_range>
          <name>High</name>
          <min>7.0</min>
          <max>10.0</max>
        </severity_range>
      </severity_class>
      <scan_run_status>Done</scan_run_status>
      <hosts>
        <count>0</count>
      </hosts>
      <closed_cves>
        <count>0</count>
      </closed_cves>
      <vulns>
        <count>0</count>
      </vulns>
      <os>
        <count>0</count>
      </os>
      <apps>
        <count>0</count>
      </apps>
      <ssl_certs>
        <count>0</count>
      </ssl_certs>
      <task id="40b6c517-ba10-4fbc-811c-1435c28be1b2">
        <name>Scan Suspect Host 192.168.56.1</name>
        <comment/>
        <target id="3006db49-89c2-435b-9d52-77e0ad414f4b">
          <trash>0</trash>
          <name>Suspect Host 192.168.56.1 2020-11-05 15:24:48.003377</name>
          <comment/>
        </target>
        <progress>-1</progress>
      </task>
      <scan>
        <task/>
      </scan>
      <timestamp>2020-11-05T09:54:48Z</timestamp>
      <scan_start/>
      <timezone>Coordinated Universal Time</timezone>
      <timezone_abbrev>UTC</timezone_abbrev>
      <result_count>0<full>0</full><filtered>0</filtered><debug><full>0</full><filtered>0</filtered></debug><hole><full>0</full><filtered>0</filtered></hole><info><full>0</full><filtered>0</filtered></info><log><full>0</full><filtered>0</filtered></log><warning><full>0</full><filtered>0</filtered></warning><false_positive><full>0</full><filtered>0</filtered></false_positive></result_count>
      <severity>
        <full>-3.0</full>
        <filtered>-3.0</filtered>
      </severity>
      <scan_end>2020-11-05T09:54:48Z</scan_end>
      <errors>
        <count>1</count>
        <error>
          <host>
            <asset asset_id=""/>
          </host>
          <port/>
          <description>Could not connect to Scanner</description>
          <nvt oid="">
            <type>nvt</type>
            <name/>
            <cvss_base/>
          </nvt>
          <scan_nvt_version/>
          <severity>-3</severity>
        </error>
      </errors>
      <report_format/>
    </report>
  </report>
  <filters id="">
    <term>apply_overrides=0 min_qod=70 first=1 rows=10 sort=name</term>
    <keywords>
      <keyword>
        <column>apply_overrides</column>
        <relation>=</relation>
        <value>0</value>
      </keyword>
      <keyword>
        <column>min_qod</column>
        <relation>=</relation>
        <value>70</value>
      </keyword>
      <keyword>
        <column>first</column>
        <relation>=</relation>
        <value>1</value>
      </keyword>
      <keyword>
        <column>rows</column>
        <relation>=</relation>
        <value>10</value>
      </keyword>
      <keyword>
        <column>sort</column>
        <relation>=</relation>
        <value>name</value>
      </keyword>
    </keywords>
  </filters>
  <sort>
    <field>name<order>ascending</order></field>
  </sort>
  <reports start="1" max="1000"/>
  <report_count>8<filtered>1</filtered><page>1</page></report_count>
</get_reports_response>
```

**Comment by BhagyashreeKoyande** [comment]
As you said, I have pasted the xml content. Please guide me further what to be done

**Comment by bjoernricks** [comment]
Your report doesn't contain any scan results. Your environment has not being set up correctly. Your scanner is not running correctly.

```xml
<description>Could not connect to Scanner</description>
```

In that case it could be possible that the PDF report format doesn't return any content and the example pdf download script crashes while trying to convert the non existing base64 content.

Please search in the forum for further advises about checking and fixing the scanner setup. There are a loooot of threads.

**Comment by BhagyashreeKoyande** [comment]
Thanks for your help @y0urself  and @bjoernricks

**Comment by bjoernricks** [comment]
I think this issue is fixed with #328
